src/System.Net.Http/src/System/Net/Http/Unix/CurlCallbacks.cs

```diff
@@ -221,13 +233,39 @@ private static bool TryParseStatusLine(HttpResponseMessage response, string resp
             // Parse first 3 characters after a space as status code
             // TODO: Parsing errors are treated as fatal. Find right behaviour
             int statusCodeLength = 3;
-            int codeIndex = responseHeader.IndexOf(' ') + 1;
-            string strStatusCode = responseHeader.Substring(codeIndex, statusCodeLength);
-            response.StatusCode = (HttpStatusCode)(int.Parse(strStatusCode));
 
-            // The remaining string after the space is the reason phrase
-            response.ReasonPhrase = responseHeader.Substring(codeIndex + statusCodeLength + 1);
+            int spaceIndex = responseHeader.IndexOf(spaceChar);
+
+            if (spaceIndex > -1)
+            {
+                int codeStartIndex = spaceIndex + 1;
+                int codeEndIndex = codeStartIndex + statusCodeLength;
+
+                if (codeEndIndex <= responseHeaderLength)
+                {
+                    string strStatusCode = responseHeader.Substring(codeStartIndex, statusCodeLength);
+                    response.StatusCode = (HttpStatusCode)(int.Parse(strStatusCode));
+
+                    int reasonPhraseIndex = codeEndIndex + 1;
 
+                    if (reasonPhraseIndex < responseHeaderLength && char.Equals(responseHeader[codeEndIndex], spaceChar))
```

Comment:
Object.Equals is almost never the right answer for value types (e.g. char).  And, yes, there are significant performance implications: each call to this as ```char.Equals(char1, char2)``` (which is really just object.Equals) is going to allocate a box for char1 and allocate a box for char2, so two allocations instead of the zero allocations you get with == on the chars.  Plus, in the former case you end up with virtual function calls, whereas in the == case you end up with just an integer comparison.